support proc-macros

```rust
#[unroll::unroll_for_loops]
fn unroll_loop(a: Vec3) -> Vec3 {
    let mut b = Vec3::default();
    for i in 0..3 {
        b[i] + 1.0;
    }
    b
}
```
```sh
cargo +nightly build                                       (base) 
   Compiling wgpu-example-runner v0.1.0 (/home/dvc/implicit/marching_rays/runner)
error: failed to run custom build command for `wgpu-example-runner v0.1.0 (/home/dvc/implicit/marching_rays/runner)`

Caused by:
  process didn't exit successfully: `/home/dvc/implicit/marching_rays/target/debug/build/wgpu-example-runner-fbbe94bc606d1aa7/build-script-build` (exit code: 1)
  --- stderr
     Compiling wgpu-example-shader v0.1.0 (/home/dvc/implicit/marching_rays/shader)
  error[E0463]: can't find crate for `unroll`
    --> shader/src/lib.rs:45:3
     |
  45 | #[unroll::unroll_for_loops]
     |   ^^^^^^ can't find crate
     |
     = note: extern location for unroll is of an unknown type: /home/dvc/implicit/marching_rays/target/release/deps/libunroll-75b6e6a2c714c829.so
     = help: file name should be lib*.rlib or *..spv

  error: aborting due to previous error

  For more information about this error, try `rustc --explain E0463`.
  error: could not compile `wgpu-example-shader`

  To learn more, run the command again with --verbose.
  Error: BuildFailed
```

I traced the issue down to some weird behavior in rustc. The gist of the issue is that when building with the spir-v backend we configure the compiler to generate `.spv` files:
https://github.com/EmbarkStudios/rust-gpu/blob/afe3e1f83fe68b35d57efe23322f896aae6aa2ae/crates/rustc_codegen_spirv/src/lib.rs#L163-L166

However, for some reason, rustc uses the same postfix and suffix (`*.spv`) when considering which libraries it can use for plugins proc-macros, instead of using the default for the system it's running on (`*.so` on Linux, like in the error above). I tried changing the lines `dll_suffix` and `dll_prefix` above to match my system (MacOS) but got a similar error:

```
     Compiling core v0.0.0 (/Users/christofer/.rustup/toolchains/nightly-2020-10-25-x86_64-apple-darwin/lib/rustlib/src/rust/library/core)
     Compiling rustc-std-workspace-core v1.99.0 (/Users/christofer/.rustup/toolchains/nightly-2020-10-25-x86_64-apple-darwin/lib/rustlib/src/rust/library/rustc-std-workspace-core)
     Compiling compiler_builtins v0.1.35
     Compiling spirv-std v0.1.0 (/Users/christofer/dev/rust-gpu/fork/crates/spirv-std)
     Compiling glam v0.9.5 (https://github.com/EmbarkStudios/glam-rs?rev=d55ff75e91aab01c4fd9808e3ca9c079dfa7315c#d55ff75e)
     Compiling proc-macro-shader v0.1.0 (/Users/christofer/dev/rust-gpu/fork/examples/shaders/proc-macro-shader)
  error[E0463]: can't find crate for `unroll`
   --> examples/shaders/proc-macro-shader/src/lib.rs:9:3
    |
  9 | #[unroll::unroll_for_loops]
    |   ^^^^^^ can't find crate

  error: aborting due to previous error

  For more information about this error, try `rustc --explain E0463`.
  error: could not compile `proc-macro-shader`
```

I'm not acquainted enough with the rustc codebase to diagnose the error any further on my own. Hopefully someone else will be able to pick this up :)

--- a/crates/rustc_codegen_spirv/src/lib.rs
+++ b/crates/rustc_codegen_spirv/src/lib.rs
@@ -231,8 +231,16 @@ impl MetadataLoader for SpirvMetadataLoader {
         link::read_metadata(path)
     }
 
-    fn get_dylib_metadata(&self, _: &Target, _: &Path) -> Result<MetadataRef, String> {
-        Err("TODO: implement get_dylib_metadata".to_string())
+    fn get_dylib_metadata(&self, target: &Target, path: &Path) -> Result<MetadataRef, String> {
+        // HACK(eddyb) this is needed to allow metadata loading for proc macros
+        // (compiled as host dylibs); perhaps it'd be better to use the `object`
+        // crate, like `rustc_codegen_cranelift` does.
+        // NOTE(eddyb) while both `::new()` and `.metadata_loader()` call `Box::new`,
+        // they only do so with ZST values, and so we don't pointlessly allocate.
+        extern crate rustc_codegen_llvm;
+        rustc_codegen_llvm::LlvmCodegenBackend::new()
+            .metadata_loader()
+            .get_dylib_metadata(target, path)
     }
 }
 

